src/isomorphism.rs

```diff
@@ -460,6 +613,91 @@ where
+                let index = n_neigh.index();
+                if st[j].out[index] > 0 && st[j].mapping[index] == end {
+                    out_count[j] += 1;
+                }
+            }
+        }
+        if out_count[0] != out_count[1] {
+            return Ok(false);
+        }
+        if g[0].is_directed() {
+            let mut out_count = [0, 0];
+            for j in graph_indices.clone() {
+                for n_neigh in g[j].neighbors_directed(nodes[j], Incoming) {
+                    let index = n_neigh.index();
+                    if st[j].out[index] > 0 && st[j].mapping[index] == end {
+                        out_count[j] += 1;
+                    }
+                }
+            }
+            if out_count[0] != out_count[1] {
+                return Ok(false);
+            }
+        }
+        // R_in
+        if g[0].is_directed() {
+            let mut in_count = [0, 0];
+            for j in graph_indices.clone() {
+                for n_neigh in g[j].neighbors(nodes[j]) {
+                    let index = n_neigh.index();
+                    if st[j].ins[index] > 0 && st[j].mapping[index] == end {
+                        in_count[j] += 1;
+                    }
+                }
+            }
+            if in_count[0] != in_count[1] {
+                return Ok(false);
+            }
+            if g[0].is_directed() {
+                let mut in_count = [0, 0];
+                for j in graph_indices.clone() {
+                    for n_neigh in g[j].neighbors_directed(nodes[j], Incoming) {
+                        let index = n_neigh.index();
+                        if st[j].ins[index] > 0 && st[j].mapping[index] == end {
+                            in_count[j] += 1;
+                        }
+                    }
+                }
+                if in_count[0] != in_count[1] {
+                    return Ok(false);
+                }
+            }
+        }
+        // R_new
+        let mut new_count = [0, 0];
+        for j in graph_indices.clone() {
+            for n_neigh in g[j].neighbors(nodes[j]) {
+                let index = n_neigh.index();
+                if st[j].out[index] == 0
+                    && (st[j].ins.is_empty() || st[j].ins[index] == 0)
+                {
+                    new_count[j] += 1;
+                }
+            }
+        }
+        if new_count[0] != new_count[1] {
+            return Ok(false);
+        }
+        if g[0].is_directed() {
+            let mut new_count = [0, 0];
+            for j in graph_indices.clone() {
+                for n_neigh in g[j].neighbors_directed(nodes[j], Incoming) {
+                    let index = n_neigh.index();
+                    if st[j].out[index] == 0 && st[j].ins[index] == 0 {
+                        new_count[j] += 1;
+                    }
+                }
+            }
+            if new_count[0] != new_count[1] {
+                return Ok(false);
+            }
+        }
```

Comment:
We repeat the same for loops multiple times, which can be tricky to maintain. I think if we only repeated it twice it would be ok, but we have it six times!

I'd refactor it making a function that takes the array and a [Direction](https://docs.rs/petgraph/0.5.1/petgraph/enum.Direction.html). To achieve the same result you have with `neighbors`, just call `neighbors_directed` with the `Outgoing` direction

Reply:
it turns out that only four loops have similar checks, i simplified them